cmd/thanos/compact.go

```diff
@@ -458,7 +459,14 @@ func runCompact(
 			return errors.Wrap(err, "sync before retention")
 		}
 
-		if err := compact.ApplyRetentionPolicyByResolution(ctx, logger, bkt, sy.Metas(), retentionByResolution, compactMetrics.blocksMarked.WithLabelValues(metadata.DeletionMarkFilename, "")); err != nil {
+		if err := compact.ApplyRetentionPolicyByResolution(
+			ctx,
+			logger,
+			bkt,
+			sy.Metas(),
+			retentionByResolution,
+			compactMetrics.blocksMarked.WithLabelValues(metadata.DeletionMarkFilename, ""),
+			noCompactMarkerFilter); err != nil {
```

Comment:
My point is: overlap is somthing out of ordinary. One block can be small enough that it will never be downsampled. Thus, if we delete it with retention policy, we would get a gap in metrics. 
I suppose that we won't be marking blocks in huge numbers, so storage usage shouldn't grow uncontrollably.

Reply:
But `no-compact-mark.json` can appear even without the user knowing: https://github.com/thanos-io/thanos/blob/main/pkg/compact/planner.go#L282-L287. I have some blocks with `no-compact-mark.json` too due to this. So, with your change they would never get deleted because they would be filtered out by the retention apply function, right?

Reply:
Applying retention could definitely be improved as you've suggested however retention only gets applied after compaction/downsampling so it shouldn't ever delete data that hasn't been compacted.

All in all, could you please remove the changes to the retention part? I believe this kind of changes should be discussed and then done separately. Then we could merge your PR.